File: BaekjoonAlgorithm/202012/BA_2667/BJAlgo_2667.py
from sys import stdin

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]


def dfs(cur_x, cur_y):
    visited[cur_x][cur_y] = 1
    for d in range(4):
        next_x = cur_x + dx[d]
        next_y = cur_y + dy[d]
        if 0 <= next_x < n and 0 <= next_y < n:
            if visited[next_x][next_y] == 0 and apartments[next_x][next_y] != 0:
                visited[cur_x][cur_y] += dfs(next_x, next_y)

    return visited[cur_x][cur_y]


n = int(stdin.readline())
apartments = []
for _ in range(n):
    apartments.append(list(map(int, stdin.readline().strip())))

visited = [[0] * n for _ in range(n)]
count = 0
result = []

for i in range(n):
    for j in range(n):
        if apartments[i][j] != 0 and visited[i][j] == 0:
            result.append(dfs(i, j))
            count += 1

print(count)
result.sort()
for r in result:
    print(r)

File: BaekjoonAlgorithm/202012/BA_2667/BJAlgo_2667.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class BJAlgo_2667 {

    private static int n;
    private static int[][] apartments;
    private static int[][] visited;

    private static int[] dx = {1, -1, 0, 0};
    private static int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        n = Integer.parseInt(br.readLine());
        apartments = new int[n][n];
        for (int i = 0; i < n; i++) {
            apartments[i] = Arrays.stream(br.readLine().split("")).mapToInt(Integer::parseInt).toArray();
        }

        visited = new int[n][n];
        int count = 0;
        ArrayList<Integer> result = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (apartments[i][j] != 0 && visited[i][j] == 0) {
                    result.add(dfs(i, j));
                    count++;
                }
            }
        }

        Collections.sort(result);
        for (int r : result) {
            sb.append(r).append("\n");
        }

        System.out.println(count + "\n" + sb);
    }

    private static int dfs(int x, int y) {
        visited[x][y] = 1;
        for (int i = 0; i < 4; i++) {
            int next_x = x + dx[i];
            int next_y = y + dy[i];

            if (next_x >= 0 && next_x < n && next_y >= 0 && next_y < n) {
                if (visited[next_x][next_y] == 0 && apartments[next_x][next_y] != 0) {
                    visited[x][y] += dfs(next_x, next_y);
                }
            }
        }
        return visited[x][y];
    }
}
